components/textbox/textbox.tsx

```diff
@@ -66,257 +67,277 @@ type Props = {
     priorityProfiles?: UserProfile[];
 };
 
-export default class Textbox extends React.PureComponent<Props> {
-    private readonly suggestionProviders: Provider[];
-    private readonly wrapper: React.RefObject<HTMLDivElement>;
-    private readonly message: React.RefObject<SuggestionBoxComponent>;
-    private readonly preview: React.RefObject<HTMLDivElement>;
-
-    static defaultProps = {
-        supportsCommands: true,
-        isRHS: false,
-        listenForMentionKeyClick: false,
-        inputComponent: AutosizeTextarea,
-        suggestionList: SuggestionList,
-    };
-
-    constructor(props: Props) {
-        super(props);
-
-        this.suggestionProviders = [];
-
-        if (props.supportsCommands) {
-            this.suggestionProviders.push(new AppCommandProvider({
-                teamId: this.props.currentTeamId,
-                channelId: this.props.channelId,
-                rootId: this.props.rootId,
+export type TextboxForwarded = {
+    getInputBox: () => any;
+    focus: () => void;
+    blur: () => void;
+}
+const TextboxComponent: React.ForwardRefRenderFunction<TextboxForwarded, Props> = ({
+    supportsCommands = true,
+    isRHS = false,
+    listenForMentionKeyClick = false,
+    inputComponent = AutosizeTextarea,
+    suggestionList = SuggestionList,
+    characterLimit,
+    handlePostError,
+    channelId,
+    currentTeamId,
+    rootId,
+    currentUserId,
+    useChannelMentions,
+    autocompleteGroups,
+    priorityProfiles,
+    actions,
+    value,
+    onChange,
+    onKeyDown,
+    onSelect,
+    onMouseUp,
+    onKeyUp,
+    onBlur,
+    onHeightChange,
+    preview,
+    emojiEnabled,
+    badConnection,
+    tabIndex,
+    onKeyPress,
+    id,
+    createMessage,
+    onComposition,
+    onPaste,
+    suggestionListPosition,
+    disabled,
+    openWhenEmpty,
+}: Props, ref) => {
+    const [suggestionProviders, updateSuggestionProviders] = React.useState<Provider[]>([]);
+    const wrapper = createRef<HTMLDivElement>();
+    const message = createRef<SuggestionBoxComponent>();
+    const previewDiv = createRef<HTMLDivElement>();
+
+    useEffect(() => {
+        const newProviders: Provider[] = [];
+
+        if (supportsCommands) {
```

Comment:
This is another thing that can be moved out into it's own

Reply:
I think doing that is actually necessary for this. As-is, this useEffect hook will re-run and recreate the providers any time any one of them might change, and that'll blow away the internal state of all the selectors. Previously, we only created them in the constructor and then we called their setProps methods to update them without messing up the internal state.

I think this shouldn't be too bad to pull out into a hook for this component though. We could also make some parts of it nicer for TS since instead of storing an array and having to loop through and find the providers that care about specific props, we could just store them as separate fields and then have the hook return an array. I'm thinking it would look something like

```ts
export function useTextboxProviders(channelId: string, rootId: string) {
    // Get everything for providers from redux like currentUserId and priorityProfiles if they
    // aren't needed by the component itself
    
    // Make the initial providers
    const atMentionProvider = useRef(new AtMentionProvider(...));
    const channelMentionProvider = useRef(new ChannelMentionProvider(...));

    // Update them when props change with each provider that cares given its own useEffect
    useEffect(() => {
        atMentionProvider.current.setProps(...);
    }, [currentUserId, channelId, ...);
    
    // Maybe there's an easier way to construct a memoized array that only changes if any
    // of the refs change, but this keeps the original behaviour where this.providers never
    // changed
    return useMemo(() => [
        atMentionProvider.current,
        channelMentionProvider.current,
    ], [atMentionProvider.current, channelMentionProvider.current]);
}
```

That'd also help to cut down on the number of props that Textbox needs too since we could switch to `useSelector` inside the hook to encapsulate the redux logic as well.

---

Certainly not necessary here, but I also wanted to mention that Providers are a big area I'd like to improve about the SuggestionBox stuff. The fact that they can't easily be connected to redux directly is part of why they're so annoying to update the props on. I've had some weird ideas about making them invisible components just to connect to redux, but that'd be a lot bigger change. Encapsulating them in a hook is actually a lot nicer and should require less changes than I expected though!